File: src/Section.cpp
#include <Section.hpp>
#include <iostream> // std::cout
#include <cmath>    // std::ceil

namespace pe_injector
{

Section::Section(const uint     &payload_size,
                 const uint32_t &file_alignment,
                 const uint32_t &section_alignment,
                 const uint     &raw_offset,
                 const uint32_t &virtual_offset,
                 const bool     &generate_data)
  : m_payload(payload_size),
    m_distribution(0, 255)

{
  // Make name a bit more random
  auto name = generate_payload(3, 33, 126);
  m_Name[4] = name[0];
  m_Name[5] = name[1];
  m_Name[6] = name[2];

  std::random_device rd;
  m_rng.seed(rd());

  m_SizeOfRawData = align(payload_size, file_alignment);

  std::cout
    << "\t [*] align ("
    << payload_size
    << ", " << file_alignment
    << ") => " << m_SizeOfRawData
  << std::endl;

  m_PointerToRawData = raw_offset;

  // it will be placed after the last section in memory
  m_VirtualAddress = align(virtual_offset, section_alignment);

  setHeader();

  if (generate_data)
  {
    m_payload = generate_payload(payload_size);
    setData();
  }
}

Section::Section(const std::vector<std::uint8_t>   &payload,
                 const uint32_t                    &file_alignment,
                 const uint32_t                    &section_alignment,
                 const uint                        &raw_offset,
                 const uint32_t                    &virtual_offset)
  : Section(payload.size(),
            file_alignment,
            section_alignment,
            raw_offset,
            virtual_offset,
            false)
{
  m_payload = payload;
  setData();
}


uint
Section::align(const uint &src,
               const uint &target)
{
  return std::ceil((double) src / (double) target) * target;
}

void
Section::setHeader()
{
  // Copy section Name
  std::copy(m_Name.begin(), m_Name.end(), m_Header.begin());

  // Copy section VirtualSize
  std::copy(m_VirtualSize.begin(),
            m_VirtualSize.end(),
            m_Header.begin() + 8);

  /* Copy uint values converting to little endian */

  m_Header.at(12) = (m_VirtualAddress >> (8*0)) & 0xff;
  m_Header.at(13) = (m_VirtualAddress >> (8*1)) & 0xff;
  m_Header.at(14) = (m_VirtualAddress >> (8*2)) & 0xff;
  m_Header.at(15) = (m_VirtualAddress >> (8*3)) & 0xff;

  m_Header.at(16) = (m_SizeOfRawData >> (8*3)) & 0xff;
  m_Header.at(17) = (m_SizeOfRawData >> (8*2)) & 0xff;
  m_Header.at(18) = (m_SizeOfRawData >> (8*1)) & 0xff;
  m_Header.at(19) = (m_SizeOfRawData >> (8*0)) & 0xff;

  m_Header.at(20) = (m_PointerToRawData >> (8*3)) & 0xff;
  m_Header.at(21) = (m_PointerToRawData >> (8*2)) & 0xff;
  m_Header.at(22) = (m_PointerToRawData >> (8*1)) & 0xff;
  m_Header.at(23) = (m_PointerToRawData >> (8*0)) & 0xff;

  std::copy(m_PointerToRelocations.begin(),
            m_PointerToRelocations.end(),
            m_Header.begin() + 24);

  std::copy(m_PointerToLinenumbers.begin(),
            m_PointerToLinenumbers.end(),
            m_Header.begin() + 28);

  std::copy(m_NumberOfRelocations.begin(),
            m_NumberOfRelocations.end(),
            m_Header.begin() + 32);

  std::copy(m_NumberOfLinenumbers.begin(),
            m_NumberOfLinenumbers.end(),
            m_Header.begin() + 34);

  std::copy(m_Characteristics.begin(),
            m_Characteristics.end(),
            m_Header.begin() + 36);
}

void
Section::setData()
{
  auto padding = gen_padding_bytes(m_SizeOfRawData - m_payload.size());

  m_Data.clear();
  m_Data.reserve(m_payload.size() + padding.size());
  m_Data.insert( m_Data.begin(), m_payload.begin(), m_payload.end() );
  m_Data.insert( m_Data.end(), padding.begin(), padding.end() );
}

void
Section::setData(const std::vector<std::uint8_t> &data)
{
  m_Data = data;
}

std::vector<std::uint8_t>
Section::generate_payload(const uint &size,
                          const uint &low,
                          const uint &high)
{
  std::vector<std::uint8_t> result(size);

  for (int n = 0; n < size; n++)
  {
    const uint number = m_distribution(m_rng);
    if (number >= low && number <= high)
    {
      result.at(n) = number;
    }
  }

  return result;
}

std::vector<std::uint8_t>
Section::gen_padding_bytes(const uint &size)
{
  std::vector<std::uint8_t> result(size);

  for (int n = 0; n < size; n++)
  {
    result.at(n) = 0;
  }

  return result;
}

} // end pe_injector namespace

File: src/Section.py
import math
import numpy as np
import random

class InjectedSection:
  """
      This class represents a section of a PE32 executable file.
      Here we define all necessary bytes to construct a new section.

      SOURCES:
      [1] https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-image_section_header
      [2]https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers
      [2] https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-image-only
  """

  def __init__ (self,
                payload_size,
                file_alignment,
                section_alignment,
                raw_offset,
                virtual_offset,
                payload=None):
    """
      Section constructor.

      Keyword arguments:
      payload_size  --
    """

    # Generate random payload if none is given
    self.payload_size = payload_size
    if (payload is None):
      self.payload = self.generate_payload(payload_size)
    else:
      print("\t[@@] Using {} bytes long payload".format(payload.shape[0]))
      self.payload = payload.tobytes()


    #######################################
    ### Section Table (Section Headers) ###
    #######################################

    # "An 8-byte, null-padded UTF-8 encoded string. If the string is exactly 8
    # characters long, there is no terminating null. " [1]
    self.Name = b'.cnn' + self.generate_payload(3, start=33, stop=126) + b'\x00'

    """
      'For executable images, this must be a multiple of FileAlignment
      from the optional header. If this is less than VirtualSize, the
      remainder of the section is zero-filled.'

      SOURCES:
      https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers

      The size (in bytes) of the section in the disk
      We might need to pad with zeros to match the FileAlignment
    """
    self.SizeOfRawData = self.align(self.payload_size, file_alignment)

    """
      "If a section contains only uninitialized data, set this member is zero."

      SOURCES:
      https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-image_section_header

      Address of section data in file, we just set it to the same
      value as the section we are replacing (then we fix that later).
    """
    self.PointerToRawData = raw_offset

    """
      'For executable images, the address of the first byte of the section
      relative to the image base when the section is loaded into memory.'

      SOURCES:
      https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers

      We will copy it from the section we are "replacing"
      It is the virtual offset from the last section when loaded to memory
      (last_section.VirtualAddress + last_section.VirtualSize)
    """
    self.VirtualAddress = self.align(virtual_offset, section_alignment) # it will be placed after the last section in memory

    # The amount of memory the loader will allocate for this section. As
    # we don't want it to be executed, we can set it to 0 and it won't take
    # space in memory
    self.VirtualSize = 0

    # "The flags that describe the characteristics of the section.".
    # (it is 4 bytes long)
    # self.characteristics = 0x40000000 	# READ
    # self.characteristics = bytes.fromhex('40000000') 	# READ
    flags = 0x40000000 | 0x00000040 # IMAGE_SCN_MEM_READ | IMAGE_SCN_CNT_INITIALIZED_DATA
    self.characteristics = flags.to_bytes(4, 'little')

    # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ #
    # @@@@@@@ UNUSED HEADER INFO @@@@@@@ #
    # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ #

    # "This is set to zero for executable images or if there are no relocations."
    # (it is 4 bytes long)
    self.PointerToRelocations = b'\x00\x00\x00\x00'

    # "This value should be zero for an image because COFF debugging information is deprecated."
    # (it is 4 bytes long)
    self.PointerToLinenumbers = b'\x00\x00\x00\x00'

    # "The number of relocation entries for the section. This is set to zero for executable images."
    # (it is 2 bytes long)
    self.NumberOfRelocations = b'\x00\x00'

    # "The number of line-number entries for the section. This value should be zero for an image because COFF debugging information is deprecated."
    # (it is 2 bytes long)
    self.NumberOfLinenumbers = b'\x00\x00'

    #############################

    # Generate data
    self.set_header()
    self.set_data()

    return

  def align(self, val_to_align, alignment):
    """
      Method to calculate section alignments.
    """
    return math.ceil(val_to_align/alignment) * alignment

  def generate_payload(self, n_bytes, start=0, stop=255):
    """
      Generate random payload with 'n_bytes' size.
    """
    return bytes([random.randint(start, stop) for b in range(n_bytes)])

  def gen_padding_bytes(self, n_bytes):
    """
      To align a section to the alignment size, we
      might need to pad it with zero bytes
    """
    return bytes(n_bytes)

  def get_full_section(self):
    """
      Get all bytes representing a section, including header
    """
    return self.header + self.data

  def get_size(self):
    """
      Get all bytes representing a section, including header
    """
    return len(self.get_full_section())

  def set_header(self):
    print(
        "\n\t [+] InjectedSection: {{ Name: {} | " \
        "VirtualSize: {} | " \
        "VirtualAddress: {} | " \
        "SizeOfRawData: {} | " \
        "PointerToRawData: {} | " \
        "PointerToRelocations: {} | " \
        "PointerToLinenumbers: {} | " \
        "NumberOfRelocations: {} | " \
        "NumberOfLinenumbers: {} | " \
        "Characteristics: {} }}\n" \
        .format(
            self.Name \
            , self.VirtualSize.to_bytes(4, 'little') \
            , self.VirtualAddress.to_bytes(4, 'little') \
            , self.SizeOfRawData.to_bytes(4, 'little') \
            , self.PointerToRawData.to_bytes(4, 'little') \
            , self.PointerToRelocations \
            , self.PointerToLinenumbers \
            , self.NumberOfRelocations \
            , self.NumberOfLinenumbers \
            , self.characteristics
        )
    )

    self.header = self.Name \
        + self.VirtualSize.to_bytes(4, 'little') \
        + self.VirtualAddress.to_bytes(4, 'little') \
        + self.SizeOfRawData.to_bytes(4, 'little') \
        + self.PointerToRawData.to_bytes(4, 'little') \
        + self.PointerToRelocations \
        + self.PointerToLinenumbers \
        + self.NumberOfRelocations \
        + self.NumberOfLinenumbers \
        + self.characteristics

    return 0

  def set_data(self):
    self.data = self.payload + self.gen_padding_bytes(self.SizeOfRawData - self.payload_size)

    return

